Move env core modules to configuration (#290)

trotzig opened the pull request:
This will allow environments to specify more things than just core
modules. We need this to fully implement Meteor support (epic story
covered in #289).

Since you can specify multiple environments to inherit from, we need to
merge some configuration options. Most importantly, we need to merge
coreModules. We probably also want to merge aliases and namedModules.
Even though we currently only use coreModules in our environment
configs, I decided to add support for merging aliases and namedModules
as well.

This is a little bit more complicated than it needs to be. That's mostly
because we're still supporting local configuration. Once we move off of
that, we can simplify things.

trotzig commented:
@rhettlivingston, @lencioni: Let me know what you think of this. I'm interested to hear your thoughts on my decision to merge certain config options, and how I ended up implementing that.

trotzig commented:
This ended up fixing #292 too I think.

lencioni commented on lib/Configuration.js:
```diff
@@ -7,6 +7,8 @@ import semver from 'semver';
 import FileUtils from './FileUtils';
 import JsModule from './JsModule';
+import meteor from './environments/meteor';
+import node from './environments/node';
```
I think it would be better if these files were called `meteorEnvironment` and `nodeEnvironment` for maximum readability.

trotzig replied:
Fair. I'll update.

trotzig replied:
eeef564

lencioni commented on lib/Configuration.js:
```diff
@@ -34,68 +37,11 @@ const DEFAULT_CONFIG = {
-    'zlib',
-  ],
-};
+const MERGABLE_CONFIG_OPTIONS = [
```
At this point in the code, I don't really understand what "mergable" means. Maybe it is worth adding a comment?

trotzig replied:
I added a comment in 8c4044e.

lencioni commented on lib/Configuration.js:
```diff
@@ -263,15 +221,39 @@ export default class Configuration {
-        pathToCurrentFile: this.pathToCurrentFile,
-        pathToImportedModule: fromFile,
-      });
+  mergedValue(values: Array<any>, key: string, fromFile: string): any {
```
This function barely uses `this`. What do you think about moving it to the top level and passing in `pathToCurrentFile` instead? That would make it a pure function.

trotzig replied:
I guess that makes sense. Fixed in bd160be.

lencioni commented on lib/Configuration.js:
```diff
@@ -263,15 +221,39 @@ export default class Configuration {
+          pathToImportedModule: fromFile,
+        });
+      }
+      if (!MERGABLE_CONFIG_OPTIONS.includes(key)) {
```
This call should maybe be moved out of the loop.

trotzig replied:
I guess, though then I'd have to potentially invoke the function in two places. I'll leave it here.

lencioni commented on lib/__tests__/Configuration-test.js:
```diff
@@ -363,14 +363,97 @@ describe('Configuration', () => {
+          },
+          {
+            aliases: {
+              bar: 'barsketball',
```
see also BASEketball https://en.wikipedia.org/wiki/BASEketball

trotzig replied:
Fantastic.

lencioni commented on lib/__tests__/Configuration-test.js:
```diff
@@ -363,14 +363,97 @@ describe('Configuration', () => {
+          },
+          {
+            aliases: () => ({
+              baz: 'bazeball',
```
see also Brännboll https://en.wikipedia.org/wiki/Br%C3%A4nnboll

trotzig replied:
Now we're talking.

lencioni commented on lib/__tests__/Configuration-test.js:
```diff
@@ -363,14 +363,97 @@ describe('Configuration', () => {
+          },
+          {
+            namedExports: () => ({
+              bar: ['barsketball'],
```
What happens if I have `foo ['something else']` here? Will it override the earlier `foo` or get added to it?

trotzig replied:
Yes. I thought about adding support for merging those arrays but I don't know if it's worth it.

lencioni commented on lib/__tests__/Configuration-test.js:
```diff
@@ -355,6 +355,103 @@ describe('Configuration', () => {
+  describe('#get namedExports', () => {
+    beforeEach(() => {
+      // TODO: replace the array-style configuration here when we have an
+      // environment that specify namedExports.
```
I don't think I really understand these TODOs.

trotzig replied:
Yeah, I was on the fence about these. The reasoning is that once we no longer have local configuration through an array, we can't test configuration merging this way. The meteor config will soon have a namedExports config, and at that time we should start using it instead of this workaround.

lencioni commented:
LGTM

lencioni commented on lib/Configuration.js:
```diff
@@ -263,15 +221,39 @@ export default class Configuration {
+          pathToImportedModule: fromFile,
+        });
+      }
+      if (MERGABLE_CONFIG_OPTIONS.indexOf(key) === -1) {
```
We could bring in a polyfill for this pretty easily. We could grab airbnb-js-shims and be set for a bunch of things, or we can bring it in directly from https://github.com/ljharb/array-includes

rhettlivingston commented:
This is awesome! Yes, I believe you've got #292 covered also. I survived the mowing and will be digging into this tomorrow.

Below is the base list of namedExports that Meteor needs to complete the picture provided by those coreModules. They can be thrown in that nifty new environments file.
```
  namedExports: {
    'meteor/accounts-base': ['AccountsClient', 'Accounts', 'AccountsServer'],
    'meteor/blaze': ['Blaze'],
    'meteor/check': ['check', 'Match'],
    'meteor/ddp-client': ['DDP'],
    'meteor/ddp-rate-limiter': ['DDPRateLimiter'],
    'meteor/ejson': ['EJSON'],
    'meteor/email': ['Email'],
    'meteor/http': ['HTTP'],
    'meteor/meteor': ['Meteor'],
    'meteor/mongo': ['Mongo'],
    'meteor/random': ['Random'],
    'meteor/reactive-var': ['ReactiveVar'],
    'meteor/session': ['Session'],
    'meteor/templating': ['Template'],
    'meteor/tracker': ['Tracker'],
  },
```

rhettlivingston commented:
Another important aspect of my thoughts for #290 involves the existing packageDependencies routine. 

packageDependencies is currently very Node specific. Thus, it seems it _should_ be part of the Node environnment. However, that might be breaking because I don't think your users have to specify the Node environment to get it right now. 

Regardless, I think the packageDependencies need to be pulled from the configuration just like namedExports. Meteor, needs both the existing packageDependencies (because it is based on Node and does now have a package.json and all that comes with it) and a routine to scan Meteor's package system.

With support for packageDependencies in the configuration, that can just be implemented there...

One question,,, I'd like to take advantage of your existing routine for scanning a module to find its named exports in Meteor's packageDependencies routine. So, will I be able to get to that from within environments/meteor.js?

trotzig commented:
@rhettlivingston: I added `namedExports` in d6340b4. 

For `packageDependencies`, we can move the default `package.json` scanning into a separate module, then make the default configuration (not the `node` one) simply use this module. And then the `meteor` environment file can override `packageDependencies` and return its own list (making use of that same module to scan `package.json`). 

There is no routine to scan for named exports. We've talked about adding one in #199, but it hasn't been implemented - mainly because of performance concerns. We haven't come up with a solution that's fast and robust enough. It should be noted however that the discussion in that issue started when this whole project was written in Ruby, and importing was done through making a call to the `importjs` CLI tool. Since then we've converted to javascript and implemented a daemon running in the background. There's a lot more room for long-running processes now, and indexing/caching etc. Also, #199 discusses adding a scanner for all js files, not just package dependencies. 

If you start working on moving `packageDependencies` to configuration, leave a line here. Otherwise I might start doing it tonight or tomorrow night.

trotzig commented:
> packageDependencies is currently very Node specific. Thus, it seems it should be part of the Node environnment. However, that might be breaking because I don't think your users have to specify the Node environment to get it right now.

I agree that's specific to Node, although our `node` environment is more for "pure" node libraries (like ImportJS for instance). For projects meant to be run in a browser, you shouldn't use the `node` environment. But you might still need `package.json`. Does that make sense?

rhettlivingston commented:
>I added namedExports in d6340b4.

Thank you! Already pulled it.

>For packageDependencies, we can move the default package.json scanning into a separate module, then make the default configuration (not the node one) simply use this module. And then the meteor environment file can override packageDependencies and return its own list (making use of that same module to scan package.json).

That sounds good.

>There is no routine to scan for named exports. We've talked about adding one in #199, but it hasn't been implemented - mainly because of performance concerns. We haven't come up with a solution that's fast and robust enough. It should be noted however that the discussion in that issue started when this whole project was written in Ruby, and importing was done through making a call to the importjs CLI tool. Since then we've converted to javascript and implemented a daemon running in the background. There's a lot more room for long-running processes now, and indexing/caching etc. Also, #199 discusses adding a scanner for all js files, not just package dependencies.

Wow, I need more sleep. Yes, I somehow crossed up your code for digging out import statements with what it would take to dig out export statements. 

As you allude, my initial case is not a full scanner. It will be limited in that I will know exactly what files in a package to go to and the file will often be a trivial one included specifically to gather all of the exports from a more complicated package into a common point. So, I think the performance impact will be limited.

So perhaps I'll create the routine to scan a module for exports. It should have a lot of similarity to your findCurrentImports / ImportStatement code for extracting imports.

>For projects meant to be run in a browser, you shouldn't use the node environment. But you might still need package.json. Does that make sense?

Actually, very much so. Meteor is an isomorphic environment. A lot of our code runs on both the client and the server because we have speculative execution on the client. So I already benefit from having this code where it is and have not found a reason to bring in the Node coreModules yet.

>If you start working on moving packageDependencies to configuration, leave a line here. Otherwise I might start doing it tonight or tomorrow night.

I'm not jumping at this aspect yet, so feel free to go for it.
